File: java/l2server/util/xml/XmlNode.java
/*
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program. If not, see <http://www.gnu.org/licenses/>.
 */

package l2server.util.xml;

import org.w3c.dom.Node;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

/**
 * @author Pere
 */
public class XmlNode
{
	private String _name;
	private Map<String, String> _attributes = new HashMap<>();
	private List<XmlNode> _children = new ArrayList<>();
	private String _text = null;

	public XmlNode(Node base)
	{
		_name = base.getNodeName();
		for (int i = 0; i < base.getAttributes().getLength(); i++)
		{
			String name = base.getAttributes().item(i).getNodeName();
			String value = base.getAttributes().item(i).getNodeValue();
			_attributes.put(name, value);
		}

		for (Node baseSubNode = base.getFirstChild(); baseSubNode != null; baseSubNode = baseSubNode.getNextSibling())
		{
			if (baseSubNode.getNodeType() == Node.ELEMENT_NODE)
			{
				_children.add(new XmlNode(baseSubNode));
			}
		}

		if (base.getFirstChild() != null)
		{
			_text = base.getFirstChild().getNodeValue();
		}
	}

	public String getName()
	{
		return _name;
	}

	public boolean hasAttributes()
	{
		return !_attributes.isEmpty();
	}

	public boolean hasAttribute(String name)
	{
		return _attributes.containsKey(name);
	}

	public boolean getBool(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"Boolean value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		if (val instanceof Boolean)
		{
			return (Boolean) val;
		}
		try
		{
			return Boolean.parseBoolean((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Boolean value required for \"" + name + "\", but found: " + val);
		}
	}

	public boolean getBool(String name, boolean deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		if (val instanceof Boolean)
		{
			return (Boolean) val;
		}
		try
		{
			return Boolean.parseBoolean((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Boolean value required for \"" + name + "\", but found: " + val);
		}
	}

	public int getInt(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"Integer value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		if (val instanceof Number)
		{
			return ((Number) val).intValue();
		}
		try
		{
			return Integer.parseInt((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Integer value required for \"" + name + "\", but found: " + val);
		}
	}

	public int getInt(String name, int deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		if (val instanceof Number)
		{
			return ((Number) val).intValue();
		}
		try
		{
			return Integer.parseInt((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Integer value required for \"" + name + "\", but found: " + val);
		}
	}

	public long getLong(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"Integer value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		if (val instanceof Number)
		{
			return ((Number) val).longValue();
		}
		try
		{
			return Long.parseLong((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Integer value required for \"" + name + "\", but found: " + val);
		}
	}

	public long getLong(String name, long deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		if (val instanceof Number)
		{
			return ((Number) val).longValue();
		}
		try
		{
			return Long.parseLong((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Long value required for \"" + name + "\", but found: " + val);
		}
	}

	public float getFloat(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"Float value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		if (val instanceof Number)
		{
			return ((Number) val).floatValue();
		}
		try
		{
			return (float) Double.parseDouble((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Float value required for \"" + name + "\", but found: " + val);
		}
	}

	public float getFloat(String name, float deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		if (val instanceof Number)
		{
			return ((Number) val).floatValue();
		}
		try
		{
			return (float) Double.parseDouble((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Float value required for \"" + name + "\", but found: " + val);
		}
	}

	public double getDouble(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"Float value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		if (val instanceof Number)
		{
			return ((Number) val).doubleValue();
		}
		try
		{
			return Double.parseDouble((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Float value required for \"" + name + "\", but found: " + val);
		}
	}

	public double getDouble(String name, double deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		if (val instanceof Number)
		{
			return ((Number) val).doubleValue();
		}
		try
		{
			return Double.parseDouble((String) val);
		}
		catch (Exception e)
		{
			throw new IllegalArgumentException("Float value required for \"" + name + "\", but found: " + val);
		}
	}

	public String getString(String name)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			throw new IllegalArgumentException(
					"String value required for \"" + name + "\", but not specified.\r\nNode: " + this);
		}
		return String.valueOf(val);
	}

	public String getString(String name, String deflt)
	{
		Object val = _attributes.get(name);
		if (val == null)
		{
			return deflt;
		}
		return String.valueOf(val);
	}

	public Map<String, String> getAttributes()
	{
		return _attributes;
	}

	public XmlNode getFirstChild()
	{
		if (_children.isEmpty())
		{
			return null;
		}

		return _children.get(0);
	}

	public List<XmlNode> getChildren()
	{
		return _children;
	}

	public String getText()
	{
		return _text;
	}

	@Override
	public String toString()
	{
		return toString(0);
	}

	private String toString(int tabDepth)
	{
		String tabs = "";
		for (int i = 0; i < tabDepth; i++)
		{
			tabs += "\t";
		}

		String result = tabs + "<" + _name;
		for (Entry<String, String> attr : _attributes.entrySet())
		{
			result += " " + attr.getKey() + "=\"" + attr.getValue() + "\"";
		}

		if (!_children.isEmpty() || _text != null && _text.length() > 0)
		{
			result += ">\r\n";
			for (XmlNode child : _children)
			{
				result += child.toString(tabDepth + 1) + "\r\n";
			}

			result += tabs + "<" + _name + ">\r\n";
		}
		else
		{
			result += " />";
		}

		return result;
	}
}

File: src/main/java/l2server/util/xml/XmlNode.kt
package l2server.util.xml

import org.w3c.dom.Node
import java.util.ArrayList
import java.util.function.Function
import kotlin.collections.HashMap
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.set

/**
 * @author Pere
 */
class XmlNode internal constructor(private val base: Node) {

	val name: String = base.nodeName

	val text: String? = base.firstChild?.nodeValue

	val firstChild: XmlNode?
		get() {
			val children = getChildren()
			return if (children.isEmpty()) {
				null
			} else children[0]
		}

	fun hasAttributes() = base.attributes.length > 0

	fun hasAttribute(name: String) = base.attributes.getNamedItem(name) != null

	private fun getAttributeValue(name: String): String? = base.attributes.getNamedItem(name)?.nodeValue

	private fun <T> parse(name: String, value: String?, expectedType: Class<T>, parseFunction: Function<String, T>): T {
		if (value == null) {
			throw IllegalArgumentException(expectedType.simpleName + " value required for \"" + name + "\", but not specified.\r\nNode: " + this)
		}
		try {
			return parseFunction.apply(value)
		} catch (e: Exception) {
			throw IllegalArgumentException(expectedType.simpleName + " value required for \"" + name + "\", but found: " + value)
		}

	}

	private fun <T> parse(name: String, value: String?, expectedType: Class<T>, parseFunction: Function<String, T>, default: T): T {
		if (value == null) {
			return default
		}
		try {
			return parseFunction.apply(value)
		} catch (e: Exception) {
			throw IllegalArgumentException(expectedType.simpleName + " value required for \"" + name + "\", but found: " + value)
		}

	}

	fun getBool(name: String): Boolean {
		return parse(name, getAttributeValue(name), Boolean::class.java, Function { java.lang.Boolean.parseBoolean(it) })
	}

	fun getBool(name: String, default: Boolean): Boolean {
		return parse(name, getAttributeValue(name), Boolean::class.java, Function { java.lang.Boolean.parseBoolean(it) }, default)
	}

	fun getInt(name: String): Int {
		return parse(name, getAttributeValue(name), Int::class.java, Function { Integer.parseInt(it) })
	}

	fun getInt(name: String, default: Int): Int {
		return parse(name, getAttributeValue(name), Int::class.java, Function { Integer.parseInt(it) }, default)
	}

	fun getLong(name: String): Long {
		return parse(name, getAttributeValue(name), Long::class.java, Function { java.lang.Long.parseLong(it) })
	}

	fun getLong(name: String, default: Long): Long {
		return parse(name, getAttributeValue(name), Long::class.java, Function { java.lang.Long.parseLong(it) }, default)
	}

	fun getFloat(name: String): Float {
		return parse(name, getAttributeValue(name), Float::class.java, Function { java.lang.Float.parseFloat(it) })
	}

	fun getFloat(name: String, default: Float): Float {
		return parse(name, getAttributeValue(name), Float::class.java, Function { java.lang.Float.parseFloat(it) }, default)
	}

	fun getDouble(name: String): Double {
		return parse(name, getAttributeValue(name), Double::class.java, Function { java.lang.Double.parseDouble(it) })
	}

	fun getDouble(name: String, default: Double): Double {
		return parse(name, getAttributeValue(name), Double::class.java, Function { java.lang.Double.parseDouble(it) }, default)
	}

	fun getString(name: String): String {
		return parse(name, getAttributeValue(name), String::class.java, Function { it })
	}

	fun getString(name: String, default: String): String {
		return parse(name, getAttributeValue(name), String::class.java, Function { it }, default)
	}

	fun getAttributes(): Map<String, String> {
		val attributes = HashMap<String, String>()
		for (i in 0 until base.attributes.length) {
			val name = base.attributes.item(i).nodeName
			val value = base.attributes.item(i).nodeValue
			attributes[name] = value
		}
		return attributes
	}

	fun getChildren(): List<XmlNode> {
		val children = ArrayList<XmlNode>()
		var baseSubNode: Node? = base.firstChild
		while (baseSubNode != null) {
			if (baseSubNode.nodeType == Node.ELEMENT_NODE) {
				val child = XmlNode(baseSubNode)
				children.add(child)
			}
			baseSubNode = baseSubNode.nextSibling
		}
		return children
	}

	fun getChildren(name: String): List<XmlNode> {
		val list = ArrayList<XmlNode>()
		for (node in getChildren()) {
			if (node.name == name) {
				list.add(node)
			}
		}

		return list
	}

	fun getChild(name: String): XmlNode? {
		val children = getChildren(name)
		return if (children.isEmpty()) {
			null
		} else children[0]

	}

	override fun toString(): String {
		return toString(0)
	}

	private fun toString(tabDepth: Int): String {
		val tabsBuilder = StringBuilder()
		for (i in 0 until tabDepth) {
			tabsBuilder.append("\t")
		}

		val tabs = tabsBuilder.toString()
		val result = StringBuilder("$tabs<$name")
		for ((key, value) in getAttributes()) {
			result.append(" ").append(key).append("=\"").append(value).append("\"")
		}

		val children = getChildren()
		if (!children.isEmpty() || text != null && text.isNotEmpty()) {
			result.append(">\r\n")
			for (child in children) {
				result.append(child.toString(tabDepth + 1)).append("\r\n")
			}

			result.append(tabs).append("<").append(name).append(">\r\n")
		} else {
			result.append(" />")
		}

		return result.toString()
	}
}
